### PR TITLE
[ADD] contract_line_invoice_by_average

### PR DESCRIPTION
**Pull request by Swapnesh-SerpentCS**
#418

This is still WIP

Cc @max3903

### CODE REVIEW

**Comment by max3903** [comment]
@Swapnesh-SerpentCS You may want to check Pedro's comment here:
https://github.com/OCA/contract/issues/418#issuecomment-554641853

**Comment by max3903** [comment]
@Swapnesh-SerpentCS Will you be able to finish this by Thursday morning US/Central?

**Comment by Swapnesh-SerpentCS** [comment]
Hello @max3903 
I have added code base on Specification https://github.com/OCA/contract/issues/418

(I will fix Travis in the last if needed)

File: contract_line_invoice_by_average/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+This module allows you to define products on a contract that needs to be invoiced by the average quantity instead of the sum.
+
+For example, you may have a contract for 3 water dispensers and have 4 sales orders in a month for a total of 15 bottles of water. Your invoice would be:
+
+    * 3 water dispensers
```
Comment (max3903) [inline contract_line_invoice_by_average/readme/DESCRIPTION.rst]:
```suggestion
* 3 water dispensers
```

File: contract_line_invoice_by_average/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+    * 5 bottles
```
Comment (max3903) [inline contract_line_invoice_by_average/readme/DESCRIPTION.rst]:
```suggestion
* 5 bottles
```

File: contract_line_invoice_by_average/readme/CONFIGURATION.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
+* Go to Invoicing > Customers > Contracts
+* Select or create a contract with a line
+* Select the product (Dispenser), the quantity (3) and the products invoiced by average quantity (Bottle of water).
+
```
Comment (max3903) [inline contract_line_invoice_by_average/readme/CONFIGURATION.rst]:
```suggestion
```

File: contract_line_invoice_by_average/models/contract_line.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    def _recurring_create_invoice(self, date_ref=False):
+        invoices = super()._recurring_create_invoice(date_ref)
+        for contract_to_invoice in self:
+            product_avg_qty_dict = {}
+            for each_contract_line in contract_to_invoice.contract_line_ids.\
+                    filtered(lambda l: l.products_invoiced_by_avg_ids):
+                for each_avg_product in each_contract_line.products_invoiced_by_avg_ids:
+                    product_avg_qty_dict.update(
+                        {each_avg_product.id: each_contract_line.quantity})
+        for each_invoice in invoices:
```
Comment (max3903) [inline contract_line_invoice_by_average/models/contract_line.py:19]: @Swapnesh-SerpentCS I get the following error when running the scheduler manually:

```
[...]
  File "/opt/odoo/src/contract/contract/models/contract.py", line 467, in cron_recurring_create_invoice
    contracts_to_invoice._recurring_create_invoice(date_ref)
  File "/opt/odoo/src/contract/contract_line_invoice_by_average/models/contract_line.py", line 19, in _recurring_create_invoice
    for each_invoice in invoices:
ValueError: <class 'TypeError'>: "'NoneType' object is not iterable" while evaluating
'model.cron_recurring_create_invoice()'
```

**Comment by max3903** [comment]
@Swapnesh-SerpentCS We demoed this to the customer and I got the design wrong. Please don't spend any additional time on this one.

**Pull request closed without merging**